File: 240612/사다리 타기/ladder-game.py
n, m = map(int, input().split())

line = []

for _ in range(m):
    a, b = map(int, input().split())
    line.append((a, a+1, b))

def link_arr(sm):

    link = {}
    for i in sm:
        if i[0] in link:
            link[i[0]].append((i[1], i[2]))
        else:
            link[i[0]] = [(i[1], i[2])]
        
        if i[1] in link:
            link[i[1]].append((i[0], i[2]))
        else:
            link[i[1]] = [(i[0], i[2])]
    return link


def next_line(link, now, sub):
    for e, u in link[now]:
        if sub < u:
            now = e
            sub = u
            return (now, sub)
    return (-1, -1)  

def move(link, length):

    count = []
    for i in range(1, n+1):
        now = i
        sub = 0
        flag = 1
        while flag:

            if now in link:
                link[now].sort(key=lambda x: x[1])

                res = next_line(link, now, sub)
                if (res[0] == -1):
                    flag = 0
                    break
                else:
                    now = res[0]
                    sub = res[1]

            else:
                flag = 0
                break
        count.append(now)
    return (count, length)

def back(cnt, sm):
    global ans
    global s 
    if (cnt >= m):
        link = link_arr(sm)
        res = move(link, len(sm))
        if res[1] == m:
            s = res[0]
        if res[0] == s:
            ans = min(ans, res[1])
            
        return

    back(cnt+1, sm+[line[cnt]])
    back(cnt+1, sm)


ans = m
s = []
back(0, [])
print(ans)

File: trail4/사다리 타기/ladder-game.java
import java.util.*;

class Input
{
    int a;
    int b;
    int c;

    Input(int a, int b, int c)
    {
        this.a = a;
        this.b = b;
        this.c = c;
    }
}

public class Main {
    public static int n, m, max, ans;
    public static int[][] arr;
    public static Input[] input;
    public static int[] full;
    public static void main(String[] args) {
       
       Scanner sc = new Scanner(System.in);
       n = sc.nextInt();
       m = sc.nextInt();
       
       input = new Input[m];
       full = new int[n+1];
       ans = m;
       for (int i = 0; i < m; i++)
       {
            int a = sc.nextInt();
            int b = sc.nextInt();

            max = Math.max(a, max);
            max = Math.max(b, max);

            input[i] = new Input(a, a+1, b);
       }

        back(0, 0);
        System.out.print(ans);

    }
    
    public static List<Input> list = new ArrayList<>();
    public static void back(int cnt, int idx)
    {
        if (cnt >= m)
        {
            move(list);
            return;
        }
        
        list.add(new Input(input[idx].a, input[idx].b, input[idx].c));
        back(cnt+1, idx+1);
        list.remove(list.size() - 1);
        back(cnt+1, idx+1);
        
    }

    public static void move(List<Input> chList)
    {
        int[] result = new int[n+1];
        chList.sort((cha, chb) -> cha.c - chb.c );

        for (int num = 1; num <= n; num++)
        {
            int now = num;
            for (int i = 0; i < chList.size(); i++)
            {
                int s = chList.get(i).a;
                int e = chList.get(i).b;
                
                if (s == now)
                {
                    now = e;
                }
                else if (e == now)
                {
                    now = s;
                }
            }
            result[num] = now;
            if (chList.size() == m)
            {
                    full[num] = now;
            }
        }

        int flag = 1;
        for (int ii = 0; ii <= n; ii++)
        {
            if (result[ii] != full[ii])
            {
                flag = 0;
                break;
            }
        }

        if (flag == 1)
        {
            ans = Math.min(ans, chList.size());
        }

    }
}